esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,914 @@
+    if (strieq(op, "if"))
+        return new CEsdlTransformOperationConditional(element, prefix);
+    if (streq(op, "set-value") || streq(op, "SetValue"))
+        return new CEsdlTransformOperationSetValue(element, prefix);
+    if (streq(op, "append-to-value") || streq(op, "AppendValue"))
+        return new CEsdlTransformOperationAppendValue(element, prefix);
+    if (streq(op, "add-value"))
+        return new CEsdlTransformOperationAddValue(element, prefix);
+    if (streq(op, "fail"))
+        return new CEsdlTransformOperationFail(element, prefix);
+    if (streq(op, "assert"))
+        return new CEsdlTransformOperationAssert(element, prefix);
+    return nullptr;
+}
+
+static IPropertyTree *getTargetPTree(IPropertyTree *tree, IXpathContext *xpathContext, const char *target)
+{
+    StringBuffer xpath(target);
+    if (xpath.length())
+    {
+        //we can use real xpath processing in the future, for now simple substitution is fine
+        StringBuffer variable;
+        xpath.replaceString("{$query}", xpathContext->getVariable("query", variable));
+        xpath.replaceString("{$method}", xpathContext->getVariable("method", variable.clear()));
+        xpath.replaceString("{$service}", xpathContext->getVariable("service", variable.clear()));
+        xpath.replaceString("{$request}", xpathContext->getVariable("request", variable.clear()));
+
+        IPropertyTree *child = tree->queryPropTree(xpath.str());  //get pointer to the write-able area
+        if (!child)
+            throw MakeStringException(-1, "EsdlCustomTransform error getting target xpath %s", xpath.str());
+        return child;
+    }
+    return tree;
+}
+static IPropertyTree *getOperationTargetPTree(MapStringToMyClass<IPropertyTree> &treeMap, IPropertyTree *currentTree, IEsdlTransformOperation &operation, IPropertyTree *tree, IXpathContext *xpathContext, const char *target)
+{
+    const char *mergedTarget = operation.queryMergedTarget();
+    if (isEmptyString(mergedTarget) || streq(mergedTarget, target))
+        return currentTree;
+    IPropertyTree *opTree = treeMap.getValue(mergedTarget);
+    if (opTree)
+        return opTree;
+    opTree = getTargetPTree(tree, xpathContext, mergedTarget);
+    if (opTree)
+        treeMap.setValue(mergedTarget, LINK(opTree));
+    return opTree;
+}
+
+class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
+{
+private:
+    IArrayOf<IEsdlTransformOperation> m_variables; //keep separate and only at top level for now
+    IArrayOf<IEsdlTransformOperation> m_operations;
+    StringAttr m_name;
+    StringAttr m_target;
+    StringBuffer m_prefix;
+
+public:
+    CEsdlCustomTransform(){}
+    CEsdlCustomTransform(IPropertyTree &tree, const char *ns_prefix) : m_prefix(ns_prefix)
+    {
+        if (m_prefix.length())
+            m_prefix.append(':');
+        else
+        {
+            const char *tag = tree.queryName();
+            if (!tag)
```

Comment:
It is possible for the root of a ptree to be unnamed.  It was like that when merging the scripts into on tree until I fixed it.  This is defensive in case the issue comes back.

Reply:
Fair enough. I'm not dealing with ptree in this sense, so it hasn't been a concern for me.